Source/WebCore/rendering/style/StyleInvalidImage.h

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "StyleGeneratedImage.h"
+
+namespace WebCore {
+
+class CSSInvalidImageValue;
+
+class StyleInvalidImage final : public StyleGeneratedImage {
```

Comment:
Does this really need a whole class? (I'm suspicious of classes that do nothing).

Reply:
This class allows us to have a representation of an invalid image that is not a null pointer. This way we can avoid adding null checks (or `std::optional::has_value` or similar) elsewhere. Also since an invalid image is used in other places in the images level 4 spec I thought it would be useful to introduce this now so it can be used if and when we implement other parts of that spec.

I think the key thing this class gives us is an object that will return a null image of size zero so both rendering and paint can get well-defined values for intrinsic size and for image contents which match the spec's notion of an invalid image.

Maybe there's another way to do that too that I'm not aware of? I think this class is a straightforward way to achieve that though I'm open to alternatives.

Reply:
I agree that CSSInvalidImage does nothing though.